Double-read of a Shelf request

```
>  DEBUG: Full request URL: http://localhost:5001/greetTyped
>  DEBUG: Request path: greetTyped
>  DEBUG: Request method: POST
>  DEBUG: Headers: {user-agent: Dart/3.12 (dart:io), connection: keep-alive, content-type: application/json, accept-encoding: gzip, content-length: 0, host: localhost:5001}
>  DEBUG: Trying to match CloudEvent function...
>  DEBUG: Content-Type: application/json
>  DEBUG: Trying structured content mode (JSON body)
>  DEBUG: Body length: 0 chars
>  Failed to parse CloudEvent for function matching: FormatException: Unexpected end of input (at character 1)
>
>  ^
>
>  DEBUG: Extracting function name from path: greetTyped
>  DEBUG: Path after removing leading slash: greetTyped
>  [ERROR] 2026-01-23 00:28:17.407914	0:00:00.005704	POST	/greetTyped
>  Bad state: The 'read' method can only be called once on a shelf.Request/shelf.Response object.
>  package:shelf/src/message.dart 165:34                             Message.readAsString
>  package:firebase_functions/src/https/callable.dart 424:33         RequestValidation.json
>  package:firebase_functions/src/https/https_namespace.dart 150:36  HttpsNamespace.onCallWithData.<fn>
>  package:firebase_functions/src/common/on_init.dart 83:25          withInit.<fn>
>  package:firebase_functions/src/server.dart 276:28                 _routeByPath
>
>  ERROR - 2026-01-23 00:28:17.423706
>  POST /greetTyped
>  Error thrown by handler.
>  Bad state: The 'read' method can only be called once on a shelf.Request/shelf.Response object.
>  package:shelf/src/message.dart 165:34                             Message.readAsString
>  package:firebase_functions/src/https/callable.dart 424:33         RequestValidation.json
>  package:firebase_functions/src/https/https_namespace.dart 150:36  HttpsNamespace.onCallWithData.<fn>
>  package:firebase_functions/src/common/on_init.dart 83:25          withInit.<fn>
>  package:firebase_functions/src/server.dart 276:28                 _routeByPath
>
i  functions: Finished "us-central1-greetTyped" in 20.058792ms
```

--- a/lib/src/server.dart
+++ b/lib/src/server.dart
@@ -352,16 +352,24 @@ Future<(Request, FirebaseFunctionDeclaration?)> _tryMatchCloudEventFunction(
       // Structured content mode - try to parse JSON body
       bodyString = await request.readAsString();
 
-      final body = jsonDecode(bodyString) as Map<String, dynamic>;
+      final Map<String, dynamic> body;
+      try {
+        body = jsonDecode(bodyString) as Map<String, dynamic>;
+      } catch (e) {
+        // Invalid JSON - not a CloudEvent request
+        return (request.change(body: bodyString), null);
+      }
 
+      final bodyType = body['type'];
+      final bodySource = body['source'];
       // Check if this is a valid CloudEvent - if not, return reconstructed request
-      if (!body.containsKey('source') || !body.containsKey('type')) {
+      if (bodyType is! String || bodySource is! String) {
         // Return the reconstructed request since we consumed the body
         return (request.change(body: bodyString), null);
       }
 
-      source = body['source'] as String;
-      type = body['type'] as String;
+      source = bodySource;
+      type = bodyType;
     }
 
     // Now we have source and type from either headers or body

--- a/test/e2e/tests/https_oncall_tests.dart
+++ b/test/e2e/tests/https_oncall_tests.dart
@@ -51,6 +51,18 @@ void runHttpsOnCallTests(FunctionsHttpClient Function() getClient) {
       );
     });
 
+    test('greet returns default name for missing "data"', () async {
+      final response = await client.post('greet', body: {});
+
+      expect(response.statusCode, equals(200));
+
+      final json = jsonDecode(response.body) as Map<String, dynamic>;
+      expect(
+        json['result'],
+        equals(<String, dynamic>{'message': 'Hello, World!'}),
+      );
+    });
+
     test('greet returns correct content type', () async {
       final response = await client.call('greet', data: {'name': 'Test'});
 
@@ -164,6 +176,12 @@ void runHttpsOnCallTests(FunctionsHttpClient Function() getClient) {
         json['result'],
         equals(<String, dynamic>{'message': 'Hello, World!'}),
       );
+    });
+
+    test('greetTyped fails when missing "data"', () async {
+      final response = await client.post('greetTyped', body: {});
+
+      expect(response.statusCode, isNot(equals(200)));
     });
 
     test('greetTyped returns correct content type', () async {

--- a/test/e2e/tests/https_onrequest_tests.dart
+++ b/test/e2e/tests/https_onrequest_tests.dart
@@ -62,6 +62,16 @@ void runHttpsOnRequestTests(
       expect(response.statusCode, equals(200));
     });
 
+    test('hello-world accepts POST requests similar to a CloudEvent', () async {
+      print('POST ${client.baseUrl}/hello-world like CloudEvent');
+      final response = await client.post(
+        'hello-world',
+        body: {'type': 0, 'source': 1},
+      );
+
+      expect(response.statusCode, equals(200));
+    });
+
     test('calling non-existent function returns 404', () async {
       print('GET ${client.baseUrl}/non-existent-function');
       final response = await client.get('non-existent-function');

